driver: PS/2: npcx: add driver support for Nuvoton npcx family

ChiHuaL opened the pull request:
The PS/2 module in npcx provides a hardware accelerator mechanism
including an 8-bit shift register, a state machine, and control logic
that handle both the incoming and outgoing data. The hardware
accelerator mechanism is shared by 4 PS/2 channels. To support it,
this CL separates the PS/2 driver into channel and controller drivers.
The controller driver is in charge of the PS/2 transaction. The channel
driver is in charge of the connection between the Zehpyr PS/2 API
interface and controller driver.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

ChiHuaL committed:
driver: PS/2: npcx: add driver support for Nuvoton npcx family

The PS/2 module in npcx provides a hardware accelerator mechanism
including an 8-bit shift register, a state machine, and control logic
that handle both the incoming and outgoing data. The hardware
accelerator mechanism is shared by 4 PS/2 channels. To support it,
this CL separates the PS/2 driver into channel and controller drivers.
The controller driver is in charge of the PS/2 transaction. The channel
driver is in charge of the connection between the Zehpyr PS/2 API
interface and controller driver.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

sjg20 commented on drivers/ps2/Kconfig.npcx:
```diff
@@ -0,0 +1,24 @@
+	bool "Nuvoton NPCX embedded controller (EC) PS2 driver"
+	depends on SOC_FAMILY_NPCX && ESPI_PERIPHERAL_8042_KBC
+	help
+	  Enable the NPCX family PS2 driver. The driver also
```
PS/2

Should describe what it supports (keyboard?) and what it does. Needs more detail

ChiHuaL replied:
Done.

sjg20 commented on drivers/ps2/ps2_npcx_channel.c:
```diff
@@ -0,0 +1,144 @@
+	/* pinmux configuration */
+	const uint8_t alts_size;
+	const struct npcx_alt *alts_list;
+	uint8_t channel_id;
```
comment for channel_id

ChiHuaL replied:
Done.

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+	 */
+	uint8_t channel_enabled_mask;
+	struct k_sem lock;
+	struct k_sem tx_sync_sem;
```
missing some member coments

ChiHuaL replied:
Done.

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+/* in 50us units */
+#define PS2_RETRY_COUNT 10000
+
+#define PS2_TRANSACTION_TIMEOUT K_MSEC(20)
```
20ms is a long time...can you add a comment as to how it is chosen?

ChiHuaL replied:
Indeed, 20ms is long. Normally, a PS/2 transaction should take about 1.1 ms. Reduce the timeout from 20 ms to 2 ms (1.1 ms + some buffer.)

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+ * Set WDAT3-0 and clear CLK3-0 in the PSOSIG register to
+ * reset the shift mechanism.
+ */
+#define NPCX_PS2_SHIFT_MECH_RESET (uint8_t) ~NPCX_PSOSIG_CLK_MASK_ALL
```
Do you need the cast? If so, can you drop the space after it?

ChiHuaL replied:
There is a compile warning if no cast. Drop the space.

albertofloyd commented:
Update codeowners file

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+	 * but both SOT and EOT are cleared when all CLKs are pull low
+	 * (due to Shift Mechanism is reset)
+	 */
+	return (IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT) |
```
Should that be || ?

Alternatively, can you use 

IS_BIT_SET(inst->PSTAT, NPCX_PSTAT_SOT | NPCX_PSTAT_EOT) ?

ChiHuaL replied:
Done by changing from "|" to "||".

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+
+	/*
+	 * ACH = 1 : CHannel 0
```
Channel

ChiHuaL replied:
Done.

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+	 */
+	active_ch = GET_FIELD(inst->PSTAT, NPCX_PSTAT_ACH);
+	active_ch = active_ch > 2 ? (active_ch - 2) : (active_ch - 1);
+	LOG_DBG("ACH: %d\n", active_ch);
```
Do you need to range-check active_ch?

ChiHuaL replied:
The active channel number is reported from the hardware module.
It should not report incorrect channel No. unless there is a bug in the hardware.
So I don't check the range of it.

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+	}
+
+	if (enable) {
+		data->channel_enabled_mask |= BIT(NPCX_PSOSIG_CLK(channel_id));
```
You use BIT(NPCX_PSOSIG_CLK(channel_id)) a lot

How about adding a function that returns the mask for a particular channel's clock?

ChiHuaL replied:
Done by adding a function to return the bit-mask.

sjg20 commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,354 @@
+
+				LOG_DBG("Recv:0x%02x", data_in);
+				callback = data->callback_isr[active_ch];
+				if (callback != NULL)
```
Is this allowed to be NULL ?

ChiHuaL replied:
The callback is assigned after the PS/2 channel is configured.
If someone enables the channel without configuring it first, the callback is NULL.

sjg20 reviewed (changes requested):
LGTM, some nits

albertofloyd commented on drivers/ps2/CMakeLists.txt:
```diff
@@ -3,3 +3,5 @@
 zephyr_library()
 
 zephyr_library_sources_ifdef(CONFIG_PS2_XEC		ps2_mchp_xec.c)
+zephyr_library_sources_ifdef(CONFIG_PS2_NPCX		ps2_npcx_channel.c)
```
This PR is missing a change into CODEOWNERS file
/drivers/ps2/*npcx*                        @ChiHual

ChiHuaL replied:
Thanks for the comment.  
I have added Nuvoton developers to the CODEOWNERS file in the 2nd commit of this PR.

ChiHuaL committed:
CODEOWNERS: PS/2: npcx: add owners for npcx PS/2 driver

Add Nuvoton developers as code owners for npcx PS/2 driver.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

WealianLiao commented on drivers/ps2/ps2_npcx_channel.c:
```diff
@@ -0,0 +1,145 @@
+	{                                                                      \
+		struct ps2_npcx_ch_data *const data = DRV_DATA(dev);           \
+									       \
+		data->ps2_ctrl = device_get_binding(DT_INST_BUS_LABEL(inst));  \
```
#36557 provides a method(DEVICE_DT_GET) to get the controller at build time. We can get the parent node by DT_PARENT.

ChiHuaL replied:
I followed PR #36557 to use DEVICE_DT_GET to get the controller device object. Please see commit #3.

ChiHuaL committed:
drivers: PS/2: npcx: Replace device_get_binding with DEVICE_DT_GET

Replace device_get_binding() with DEVICE_DT_GET to obtain the PS/2
controller and clock control device objects. It helps to improve the
efficiency for driver initialization as DEVICE_DT_GET is processed
at the link time rather than run time.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

WealianLiao commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,372 @@
+	const struct ps2_npcx_ctrl_config *const config = DRV_CONFIG(dev);
+	struct ps2_npcx_ctrl_data *const data = DRV_DATA(dev);
+	struct ps2_reg *const inst = HAL_PS2_INSTANCE(dev);
+	const struct device *clk_dev = device_get_binding(NPCX_CLK_CTRL_NAME);
```
#36557 just merged. Here should modify after rebase. `DEVICE_DT_GET(NPCX_CLK_CTRL_NODE);`

ChiHuaL replied:
Done.

keith-zephyr commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,372 @@
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
```
This path leaves the interrupt disabled and still owns the semaphore.

ChiHuaL replied:
Done by giving the semaphore back before return.

keith-zephyr replied:
What about the interrupt?  It is disabled on if channel_id is invalid, but never enabled.

ChiHuaL replied:
Yes, it should be re-enabled before return. I have fixed it. Thanks for catching it.

keith-zephyr commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,372 @@
+
+	if (channel_id >= NPCX_PS2_CH_COUNT) {
+		LOG_ERR("unexpected channel ID: %d", channel_id);
+		return -EINVAL;
```
Need to give back the semaphore or do this check before k_sem_take().

ChiHuaL replied:
Done.

keith-zephyr commented on drivers/ps2/ps2_npcx_controller.c:
```diff
@@ -0,0 +1,372 @@
+	}
+
+	if (callback_isr == NULL) {
+		return -EINVAL;
```
Same comment here - move this outside the critical section.

ChiHuaL replied:
Done.

mbolivar-nordic commented on drivers/ps2/ps2_npcx_channel.c:
```diff
@@ -0,0 +1,117 @@
+		.channel_id = DT_INST_PROP(inst, channel),                     \
+		.alts_size = ARRAY_SIZE(ps2_channel_alts##inst),               \
+		.alts_list = ps2_channel_alts##inst,                           \
+		.ps2_ctrl = DEVICE_DT_GET(DT_PARENT(DT_DRV_INST(inst))),       \
```
I'm not seeing a call to device_is_ready to check the status of the ps2_ctrl device before using it. Where is it?

ChiHuaL replied:
Thanks for pointing out this. 
I added the checking in the function ps2_npcx_channel_init.

ChiHuaL commented on drivers/ps2/ps2_npcx_channel.c:
```diff
@@ -0,0 +1,117 @@
+									       \
+	DEVICE_DT_INST_DEFINE(inst, ps2_npcx_channel_init, NULL, NULL,         \
+			      &ps2_npcx_ch_cfg_##inst, POST_KERNEL,            \
+			      CONFIG_PS2_CHANNEL_INIT_PRIORITY,                \
```
Done.
